### PR TITLE
enable 3d parallism

### PR DESCRIPTION
**Pull request by ruisizhang123**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -1035,6 +1035,12 @@ class simplefsdp:
     enable_bucket = False
     enable_reorder = False
 
+    transformer_block_bucket = False
+
+    fsdp_world_size = -1
+    tp_world_size = -1
+    pp_world_size = -1
```
Comment (tianyu-l) [inline torch/_inductor/config.py]: Let's call them `fsdp_degree`, `tp_degree`, `pp_degree`

File: torch/_inductor/config.py [hunk 2]
```diff
@@ -1035,6 +1035,12 @@ class simplefsdp:
     enable_bucket = False
```
Comment (tianyu-l) [inline torch/_inductor/config.py]: remove

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -1599,20 +1599,23 @@ def __init__(self, nodes: List[ir.Operation]) -> None:
 
         front_node = None
         if config.simplefsdp.enable_bucket:
```
Comment (tianyu-l) [inline torch/_inductor/scheduler.py]: remove

File: torch/_inductor/scheduler.py [hunk 4]
```diff
@@ -1599,20 +1599,23 @@ def __init__(self, nodes: List[ir.Operation]) -> None:
...
-            # get the first compute node w/o AG in backward graph
-            front_node = reorder.get_front_node(self.nodes)
-            # bucket all gather
-            self.nodes = transformer_block_bucket.bucket_all_gather_by_block(
-                self, self.nodes
-            )
+            if config.simplefsdp.transformer_block_bucket:
+                if config.simplefsdp.pp_world_size < 0:
```
Comment (tianyu-l) [inline torch/_inductor/scheduler.py]: add a comment on why we don't do this with PP

File: torch/_inductor/simple_fsdp/bucket.py [hunk 5]
```diff
@@ -4,7 +4,7 @@
 import torch
 import torch.distributed as dist
 
-from .. import ir, scheduler
+from .. import ir, scheduler, config
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/bucket.py]: let's not import all of `config` if you only need `config.simplefsdp`
can we do something like `from ..config.simplefsdp import fsdp_degree`

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 6]
```diff
@@ -59,12 +59,18 @@ def all_gather_copy_in_meta(
     device: torch.device,
     simplefsdp: bool = False,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
+    rank = rank%world_size
```
Comment (tianyu-l) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py]: lint?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 7]
```diff
@@ -80,6 +86,7 @@ def all_gather_copy_in_cuda(
...
```
Comment (tianyu-l) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py]: ditto

File: torch/_inductor/simple_fsdp/utils.py [hunk 8]
```diff
@@ -52,32 +52,59 @@ def compute_node_users(
...
+    ir_node_origins = list(getattr(ir_node, "origins", None))
+
+    if len(ir_node_origins) == 0:
+        # bucketed AG and RS doesn't have origins, but they are created by FSDP
+        is_fsdp = True
+
+    for n in ir_node_origins:
+        meta_data = n.meta.get("stack_trace", {})  
+        # TODO(ruisizhang123): hack to get FSDP node (the FSDP AG/RS are created from torch_spmd)
+        if 'torch_spmd' in meta_data:
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/utils.py]: maybe let's use `parametrization` as a more fundamental characterization of when FSDP happens. We may need to revisit it if later we want to also enable DDP/HSDP.

**Review by tianyu-l — APPROVED** [review]
looks good, please address comments before merge

File: torch/_inductor/simple_fsdp/bucket.py (hunk 5 above)
Comment (ruisizhang123) [reply torch/_inductor/simple_fsdp/bucket.py]: I was having problem importing `..config.simple fsdp`, I changed it to `from ..config import simplefsdp`, and call `simplefsdp.fsdp_degree`
`

File: torch/_inductor/config.py (hunk 2 above)
Comment (ruisizhang123) [reply torch/_inductor/config.py]: We cannot remove `enable_bucket`, because it was also called [LINK](https://github.com/tianyu-l/pytorch_intern24/blob/211847de3c1c3d6cbd299e14a001b794eabf2a2d/torch/_inductor/ir.py#L5994) [LINK](https://github.com/tianyu-l/pytorch_intern24/blob/211847de3c1c3d6cbd299e14a001b794eabf2a2d/torch/_inductor/ir.py#L7097).


We can disable user API, and do an additional `enable_bucket` config in the scheduler.py

Comment (ruisizhang123) [reply torch/_inductor/config.py]: I updated it as bucket_mode, as we discussed offline

File: torch/_inductor/simple_fsdp/utils.py [hunk 9]
```diff
@@ -52,32 +52,57 @@ def compute_node_users(
 
     return inverse_users, node_users
 
+
+def _check_ir_node_fsdp(ir_node: "ir.Operation") -> bool:
+    """
+    Determine if the AG/RS node is for FSDP or TP
+    """
+    if config.simplefsdp.tp_degree < 0:
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/utils.py:60]: ditto we can import `simplefsdp` only
